File: engine/src/main/java/com/example/mypixel/model/node/S3InputNode.java
package com.example.mypixel.model.node;

import com.example.mypixel.exception.InvalidNodeParameter;
import com.example.mypixel.model.Parameter;
import com.example.mypixel.model.ParameterType;
import com.example.mypixel.model.ParamsMap;
import com.example.mypixel.model.Widget;
import com.example.mypixel.service.FileHelper;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.extern.slf4j.Slf4j;
import org.springframework.lang.NonNull;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.AwsCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.S3ClientBuilder;
import software.amazon.awssdk.services.s3.model.GetObjectRequest;
import software.amazon.awssdk.services.s3.model.ListObjectsV2Request;
import software.amazon.awssdk.services.s3.model.ListObjectsV2Response;
import software.amazon.awssdk.services.s3.model.S3Object;

import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.util.HashSet;
import java.util.List;
import java.util.Map;


@MyPixelNode("S3Input")
@Slf4j
public class S3InputNode extends Node {

    @JsonCreator
    public S3InputNode(
            @JsonProperty("id") @NonNull Long id,
            @JsonProperty("type") @NonNull String type,
            @JsonProperty("inputs") Map<String, Object> inputs) {
        super(id, type, inputs);
    }

    @Override
    public Map<String, Parameter> getInputTypes() {
        return ParamsMap.of(
                "access_key_id", Parameter.builder()
                        .type(ParameterType.STRING)
                        .required(true)
                        .widget(Widget.INPUT)
                        .build(),
                "secret_access_key", Parameter.builder()
                        .type(ParameterType.STRING)
                        .required(true)
                        .widget(Widget.INPUT)
                        .build(),
                "region", Parameter.builder()
                        .type(ParameterType.STRING)
                        .required(true)
                        .widget(Widget.INPUT)
                        .build(),
                "bucket", Parameter.builder()
                        .type(ParameterType.STRING)
                        .required(true)
                        .widget(Widget.INPUT)
                        .build(),
                "endpoint", Parameter.builder()
                        .type(ParameterType.STRING)
                        .required(false)
                        .widget(Widget.INPUT)
                        .build()
        );
    }

    @Override
    public Map<String, Object> getDefaultInputs() {
        return ParamsMap.of(
                "access_key_id", "",
                "secret_access_key", "",
                "region", "",
                "bucket", ""
        );
    }

    @Override
    public Map<String, Parameter> getOutputTypes() {
        return ParamsMap.of("files", Parameter.required(ParameterType.FILEPATH_ARRAY));
    }

    @Override
    public Map<String, String> getDisplayInfo() {
        return ParamsMap.of(
                "category", "IO",
                "description", "Load files from S3",
                "color", "#AED581",
                "icon", "S3Icon"
        );
    }

    @Override
    public Map<String, Object> exec() {
        Map<String, Object> outputs;

        String accessKey = (String) inputs.get("access_key_id");
        String secretKey = (String) inputs.get("secret_access_key");
        String regionName = (String) inputs.get("region");
        String bucket = (String) inputs.get("bucket");
        String endpoint = (String) inputs.get("endpoint");

        HashSet<String> files = new HashSet<>();

        AwsCredentials credentials = AwsBasicCredentials.create(accessKey, secretKey);

        S3ClientBuilder clientBuilder = S3Client.builder()
                .region(Region.of(regionName))
                .credentialsProvider(StaticCredentialsProvider.create(credentials));

        if (endpoint != null && !endpoint.isEmpty()) {
            clientBuilder.endpointOverride(URI.create(endpoint))
                    .forcePathStyle(true);
        }

        try (S3Client s3Client = clientBuilder.build()) {
            ListObjectsV2Request listObjectsV2Request = ListObjectsV2Request.builder()
                    .bucket(bucket)
                    .build();
            ListObjectsV2Response listObjectsV2Response = s3Client.listObjectsV2(listObjectsV2Request);

            List<S3Object> contents = listObjectsV2Response.contents();

            for (S3Object file: contents) {
                String filename = file.key();
                log.debug("Loading file from S3: {}", filename);
                InputStream in = s3Client
                        .getObject(GetObjectRequest.builder().bucket(bucket).key(filename).build());

                files.add(FileHelper.storeToTaskContext(taskId, id, in, filename));
                try {
                    in.close();
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }
        }

        outputs = Map.of("files", files);
        return outputs;
    }

    @Override
    public void validate() {
        if (inputs.get("access_key_id").toString().isEmpty()) {
            throw new InvalidNodeParameter("Access key ID cannot be blank.");
        }
        if (inputs.get("secret_access_key").toString().isEmpty()) {
            throw new InvalidNodeParameter("Secret cannot be blank.");
        }
        if (inputs.get("region").toString().isEmpty()) {
            throw new InvalidNodeParameter("Region cannot be blank.");
        }
        if (inputs.get("bucket").toString().isEmpty()) {
            throw new InvalidNodeParameter("Bucket cannot be blank.");
        }
    }
}

File: node/nodes/s3_input_node.py
import os
import boto3
from typing import Dict, Any
import logging

from node import Node

logger = logging.getLogger(__name__)


class S3InputNode(Node):
    node_type = "S3Input"

    def get_input_types(self) -> Dict[str, Dict[str, Any]]:
        return {
            "access_key_id": {
                "type": "STRING",
                "required": True,
                "widget": "INPUT",
                "default": ""
            },
            "secret_access_key": {
                "type": "STRING",
                "required": True,
                "widget": "INPUT",
                "default": ""
            },
            "region": {
                "type": "STRING",
                "required": True,
                "widget": "INPUT",
                "default": ""
            },
            "bucket": {
                "type": "STRING",
                "required": True,
                "widget": "INPUT",
                "default": ""
            },
            "endpoint": {
                "type": "STRING",
                "required": False,
                "widget": "INPUT",
                "default": ""
            }
        }

    def get_output_types(self) -> Dict[str, Dict[str, Any]]:
        return {
            "files": {
                "type": "FILEPATH_ARRAY",
                "required": True
            }
        }

    def get_display_info(self) -> Dict[str, str]:
        return {
            "category": "IO",
            "description": "Load files from S3",
            "color": "#AED581",
            "icon": "S3Icon"
        }

    def exec(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
        access_key = inputs.get("access_key_id", "")
        secret_key = inputs.get("secret_access_key", "")
        region_name = inputs.get("region", "")
        bucket = inputs.get("bucket", "")
        endpoint = inputs.get("endpoint", "")

        files = set()

        session = boto3.Session(
            aws_access_key_id=access_key,
            aws_secret_access_key=secret_key,
            region_name=region_name
        )

        s3_config = {}
        if endpoint and endpoint.strip():
            s3_config['endpoint_url'] = endpoint
            s3_config['use_ssl'] = endpoint.startswith('https')
            s3_config['verify'] = False  # For testing only

        s3_client = session.client('s3', **s3_config)

        response = s3_client.list_objects_v2(Bucket=bucket)

        if 'Contents' in response:
            for obj in response['Contents']:
                filename = obj['Key']
                logger.debug(f"Loading file from S3: {filename}")

                response = s3_client.get_object(Bucket=bucket, Key=filename)
                content = response['Body'].read()

                from storage_client import StorageClient

                task_id = inputs.get("meta", {}).get("taskId")
                node_id = inputs.get("meta", {}).get("nodeId")

                if task_id and node_id:
                    temp_file_path = f"/tmp/{filename}"
                    with open(temp_file_path, 'wb') as f:
                        f.write(content)

                    file_path = StorageClient.store_to_task(
                        task_id=task_id,
                        node_id=node_id,
                        file_path=temp_file_path,
                        target=filename
                    )

                    files.add(file_path)

                    os.remove(temp_file_path)

        task_id = inputs.get("meta", {}).get("taskId")
        node_id = inputs.get("meta", {}).get("nodeId")

        output_files = []

        for file in files:
            output_files.append(StorageClient.store_from_workspace_to_task(task_id, node_id, file))

        return {"files": output_files}

    def validate(self, inputs: Dict[str, Any]) -> None:
        if not inputs.get("access_key_id", ""):
            raise ValueError("Access key ID cannot be blank.")
        if not inputs.get("secret_access_key", ""):
            raise ValueError("Secret cannot be blank.")
        if not inputs.get("region", ""):
            raise ValueError("Region cannot be blank.")
        if not inputs.get("bucket", ""):
            raise ValueError("Bucket cannot be blank.")